File: backend-beeyond/src/main/java/at/htl/beeyond/resource/TemplateResource.java
package at.htl.beeyond.resource;

import at.htl.beeyond.dto.TemplateDto;
import at.htl.beeyond.entity.Template;
import at.htl.beeyond.entity.User;

import javax.annotation.security.RolesAllowed;
import javax.transaction.Transactional;
import javax.validation.Valid;
import javax.ws.rs.*;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.Response.Status;
import javax.ws.rs.core.SecurityContext;

@Path("/template")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class TemplateResource {

    @GET
    @RolesAllowed({"student", "teacher"})
    @Transactional
    public Response getAll() {
        return Response.ok(Template.getDtos()).build();
    }

    @GET
    @Path("/{id}")
    @RolesAllowed({"student", "teacher"})
    @Transactional
    public Response getById(@PathParam("id") Long id) {
        Template template = Template.findById(id);
        if (template == null) {
            return Response.status(404).build();
        }
        return Response.ok(Template.getDto(id)).build();
    }

    @POST
    @RolesAllowed("teacher")
    @Transactional
    public Response create(@Context SecurityContext sc, @Valid TemplateDto templateDto) {
        User owner = User.find("name", sc.getUserPrincipal().getName()).firstResult();
        Template template = templateDto.map(owner);
        template.persist();

        return Response.noContent().build();
    }

    @GET
    @Path("/{id}")
    @RolesAllowed({"student", "teacher"})
    @Transactional
    public Response getTemplateById(@PathParam("id") Long id) {
        Template template = Template.findById(id);
        if (template == null) {
            return Response.status(Status.NOT_FOUND).build();
        }

        TemplateDto templateDto = TemplateDto.map(template);
        return Response.ok(templateDto).build();
    }

    @DELETE
    @Path("/{id}")
    @RolesAllowed("teacher")
    @Transactional
    public Response delete(@PathParam("id") Long id) {
        Template template = Template.findById(id);
        if (template == null) {
            return Response.status(404).build();
        }

        template.delete();
        return Response.noContent().build();
    }
}

File: backend-beeyond/src/main/java/at/htl/beeyond/resource/TemplateResource.kt
package at.htl.beeyond.resource

import at.htl.beeyond.dto.TemplateDto
import at.htl.beeyond.entity.Template
import at.htl.beeyond.entity.User
import io.quarkus.hibernate.orm.panache.PanacheEntityBase
import javax.annotation.security.RolesAllowed
import javax.transaction.Transactional
import javax.validation.Valid
import javax.ws.rs.*
import javax.ws.rs.core.*
import kotlin.streams.toList

@Path("/template")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
class TemplateResource {

    @GET
    @Transactional
    @RolesAllowed("student", "teacher")
    fun getAllTemplates(@Context ctx: SecurityContext): Response {
        return if (ctx.isUserInRole("teacher")) {
            Response.ok(Template.streamAll<Template>().map { TemplateDto(it) }.toList()).build()
        } else {
            Response.ok(Template.streamAll<Template>().map { TemplateDto(it) }.filter { !it.deleted!! }.toList()).build()
        }
    }

    @POST
    @RolesAllowed("teacher")
    @Transactional
    fun createTemplate(
            @Context sc: SecurityContext,
            @Context uriInfo: UriInfo,
            @Valid templateDto: TemplateDto?
    ): Response {
        val owner = User.find<PanacheEntityBase>("name", sc.userPrincipal.name).firstResult<User>()
        val template = Template(templateDto, owner)
        template.persist()

        val uri = uriInfo.absolutePathBuilder.path(template.id.toString()).build()
        return Response.created(uri).build()
    }

    @GET
    @Path("/{id}")
    @RolesAllowed("student", "teacher")
    @Transactional
    fun getTemplateById(@PathParam("id") id: Long?, @Context ctx: SecurityContext): Response {
        val template = Template.findById<Template>(id)
                ?: return Response.status(Response.Status.NOT_FOUND).build()

        if (ctx.isUserInRole("student") && template.deleted) {
            return Response.status(Response.Status.NOT_FOUND).build()
        }

        val templateDto = TemplateDto(template)
        return Response.ok(templateDto).build()
    }

    @DELETE
    @Path("/{id}")
    @RolesAllowed("teacher")
    @Transactional
    fun delete(@PathParam("id") id: Long?): Response {
        val template = Template.findById<Template>(id)
                ?: return Response.status(Response.Status.NOT_FOUND).build()

        template.deleted = true
        return Response.noContent().build()
    }
}
